### ISSUE TITLE
Transient custom resource failures during CRD storage version update

### ISSUE THREAD

**Issue by jpbetz** [kubernetes/kubernetes#77752]
Found this last night while adding storage revision integration tests CRDs.

When a CRD is updated, the rest storage for the [old version of the CRD is torn down](https://github.com/kubernetes/kubernetes/blob/07e3b7cdcac30c513d40f5d16f43bc4d2f869f67/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L374), which also immediately closes the etcd client for that storage. When this happens, any in-flight requests being handled by that etcd client either fail with a "context canceled" or a "grpc: the client connection is closing" error.

CRDs use a 1 write + N reads concurrency model using [contentStorageLock](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L80) for writes and [customStorage atomic.Value](https://github.com/kubernetes/kubernetes/blob/07e3b7cdcac30c513d40f5d16f43bc4d2f869f67/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go#L85) for reads. This ensures mutable operations on CRDs are safe, and ensures that once a write is complete there there is an atomic switch to the new CRD state for readers. But there is nothing to ensure that all readers are done using a REST storage before it is torn down.

This is easy to replicate via a test:

- Create a CRD with two versions (v1beta1, v1beta2), where the storage version is v1beta1
- Create a v1beta1 CR
- Update the CRD, setting the storage version to v1beta2
- Create another v1beta1 CR

The last create will typically fail with either:
- `rpc error: code = Canceled desc = grpc: the client connection is closing`
- `context canceled`

As an experiment, I added a grace period to closing the etcd client owned by the rest storage:

```go
go func() {                                                                                                                                                                               
  select {                                                                                                                                                                            
    case <-time.After(5 * time.Second):                                                                                                                                                 
      // Grace period for in-flight requests                                                                                                                                      
      client.Close()                                                                                                                                                              
    }
}()
```

And was unable to replicate the error, but this doesn't feel like a comprehensive fix.  If we want to prevent client requests from failing, we need ensure all readers are done with a rest storage before tearing it down.

cc @liggitt @roycaihw @sttts

**Comment by liggitt** [comment]
duplicate of https://github.com/kubernetes/kubernetes/issues/68508, right?

**Comment by jpbetz** [comment]
@liggitt Yes, that's it. Closing this as s duplicate of https://github.com/kubernetes/kubernetes/issues/68508
